core/ir/isa_regdeps/encoding_common.h

```diff
@@ -0,0 +1,126 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef _REGDEPS_ENCODING_COMMON_H_
+#define _REGDEPS_ENCODING_COMMON_H_
+
+/* Here we describe the encoding scheme for the Synthetic ISA that is enforced in decode.c
+ * and encode.c.
+ *
+ * Encoded instructions are 4 byte aligned.
+ *
+ * All instruction encodings begin with the following 4 header bytes, which follow this
+ * scheme:
+ * |----------------------| |--| |----| |----|
+ * 31..               ..10  9,8   7..4   3..0
+ *        category         eflags #src   #dst
+ *
+ * - 22 bits, category: it's a high level representation of the opcode of an instruction.
+ *   Each bit represents one category following #dr_instr_category_t.  Note that an
+ *   instruction can belong to more than one category, hence multiple bits can be set;
+ * - 2 bits, eflags: most significant bit set to 1 indicates the instruction reads at
+ *   least one arithmetic flag; least significant bit set to 1 indicates the instruction
+ *   writes at least one arithmetic flag;
+ * - 4 bits, #src: number of source operands (read) that are registers.  Registers used in
+ *   memory reference operands of the instruction we are encoding (regardless of whether
+ *   they are source or destination operands) are considered as source operands in the
+ *   encoded instruction because they are always read;
+ * - 4 bits, #dst: number of destination operands (written) that are registers.
+ *
+ * We assume these encoded values to be little-endian.  Note that we are only interested
+ * in register dependencies, hence operands that are not registers, such as immediates or
+ * memory references, are not present.
+ *
+ * Following the 4 header bytes are the bytes for the operation size and for the encoding
+ * of register operands, if any are present.
+ *
+ * The first byte contains the operation size encoded as an OPSZ_ enum value.  The
+ * operation size is the size of the largest source operand, regardless of it being a
+ * register, a memory reference, or an immediate.
+ *
+ * Following the operation size are the register operand IDs.  Each register operand is 1
+ * byte.  The destination operands go first, followed by the source operands.  An
+ * instruction can have up to 8 operands (sources + destinations).  Note that, because of
```

Comment:
Please make sure there's a check/warning if some instr happens to have >8 reg operands (e.g. we missed some x86 corner case) so we'll know.

Reply:
Done in encoding.